include/pybind11/pytypes.h

```diff
@@ -1829,7 +1829,7 @@ class capsule : public object {
             // guard if destructor called while err indicator is set
             error_scope error_guard;
             auto destructor = reinterpret_cast<void (*)(void *)>(PyCapsule_GetContext(o));
-            if (PyErr_Occurred()) {
+            if (destructor == nullptr && PyErr_Occurred()) {
```

Comment:
Main thing here is PyErr_Occurred() is a much more expensive call, especially on the common path so might as well avoid doing so if possible

Reply:
Hm ... but what if the error indicator is set (for whatever reason)? We're leaving the process in an invalid state, follow-on Python C API calls may behave in confusing ways (painful debugging memories come back).

If the Python documentation guarantees that the error indicator is not set unless `destructor == nullptr`, this change is fine.

Looking: https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_GetContext
________ 
Return the current context stored in the capsule. On failure, set an exception and return NULL.

It is legal for a capsule to have a NULL context. This makes a NULL return code somewhat ambiguous; use [PyCapsule_IsValid()](https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_IsValid) or [PyErr_Occurred()](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_Occurred) to disambiguate.
________

To me it sounds like:

* If the error indicator is set the return value is certainly nullptr.
* But knowing that the return value is nullptr does not tell us anything about the state of the error indicator.

So checking the error indicator is mandatory, unless we want to risk leaving the process in an invalid state, which I don't think is a responsible thing to do, even it this was more than a minor optimization.

Reply:
@rwgk The documentation and the [reference implementation](https://github.com/python/cpython/blob/f4095e53ab708d95e019c909d5928502775ba68f/Objects/capsule.c#L122) seem to guarantee that the error indicator is not set unless it returns a nullptr. 

A better way to state this is the following. Is it ever valid to set the ErrorIndicator and return a non-nullptr from this method? That is there any case where non-null and the error indicator is set (for the return to this function. Ignoring it could be set to an err due leakage from some other part of the code.

The answer is no, because if it returned any non-null ptr, then an error did not occur. The error only occurs if the capsule is [PyCapsule_IsValid()](https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_IsValid) returns false internally, it therefore unable to access the members of the pycapsule, and then return null. Therefore, this function only sets the error indicator IFF it returns a nullptr. The error guard guarantees that we the error indicator is not set. 

Therefore, destructor being non-null strictly implies that no error has occurred.

Reply:
I agree, sorry for getting this wrong. I didn't look at the implementation yesterday, from that it's totally clear.